rfcs/0169-feature-parameter-names.md

```diff
@@ -0,0 +1,572 @@
+---
+feature: feature_parameter_names
+start-date: 2024-01-10
+author: Dmitry Bogatov
+co-authors: (find a buddy later to help out with the RFC)
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: (will contain links to implementation PRs)
+---
+
+# Summary
+[summary]: #summary
+
+Nixpkgs has mechanisms to exports compile-time options of the upstream build
+system through the named parameters of derivation function, but the names of
+these parameters are inconsistent from package to package and from feature to
+feature.
+
+# Motivation
+[motivation]: #motivation
+
+Developer interface for building environment with specific features enabled and
+disabled is more complicated and and involves more "you just have to know it"
+compared to already existing systems.
+
+1. Knowing that package has a feature parameter (out of knowing the upstream)
+   does not mechanically translates to the name of feature parameter, it can be
+   named in multiple ways.
```

Comment:
Doesn't `.overrideAttrs (old: { cmakeFlags = old.cmakeFlags or [] ++ [ "..." ] })` already do this in a fairly standard way? I don't think there's even a need to expose cmake flags further.

Reply:
Not everything uses cmake. There are multiple build systems around, and projects even switch them sometimes. And each project names corresponding flag slightly differently.

The whole point of distribution is to abstract away idiosyncrasies of the upstream. If we say that anybody who wants non-default configuration has to do their own research, a lot of effort will be duplicated.

Reply:
Then I'm not sure if the motivational point of being able to re-use upstream flags really works out.

If you abstract over something, you need to deal with changes when they happen underneath. This works out if upstream never changes any build flags, but we cannot guarantee that. And different build systems might not even support the same naming conventions.

So if upstream ever renames a flag, say from `with_foo` to `with_foo2`, to make the abstraction work, the Nix code has to implement support for both flags and deprecate the older one. Otherwise you risk breaking users `.override`'s.

While this is doable, I don't think package authors should be bothered with this. If the upstream flag changes, the package implementation should just transparently switch to the new name without requiring any changes at `.override` call sites.

So this is why I think this motivational point should be dropped: We shouldn't be required to re-use upstream flags.

Reply:
I see your point. I'll reword.

My idea is that if I happen to know that emacs has optional dependency on alsa, and I happend to know
that it is exported by nixpkgs, than I can write this:

```nix
emacs.override { with_alsa = false }
```

No need to remember whether it is `supportAlsa` or `withAlsaLib`.

Reply:
How about establishing `.override { alsa = null; }` as a pattern for that instead? This way we save a flag.

In the implementation this would look like

```nix
{
  stdenv,
  alsa,
}:
stdenv.mkDerivation {
  
  configureFlags = [
    (lib.optional (alsa != null) "--with-alsa")

    # Or with a more convenient function
    (lib.withPackageFeature alsa "alsa")
  ];

}
```

Reply:
This is what we used to do in Nixpkgs but I'm not a fan of this. A couple of reasons off the top of my head:

* it only applies to `with` parameters. `enable` or `conf` parameters cannot be covered by this as they aren't usually connected to any one package
* You cannot sensibly disable support for something by default; you'd have to override the thing with null at callPackage site. Apply the "ffmpeg test" if you want to see what I mean.
* Conflates existance of the dep with enablement of dep
* Overriding a dep with null could cause exceptions in the drv because it assumes the dep is an attrset (e.g. `versionAtLeast dep.version "1.0"` would throw if dep was null)

It'd be great if we could have these reasons (and likely more) to the RFC text as an alternative and for historical context as this was the status quo at some point.

Reply:
I'll include your nice list into description of `alsa = null` alternative, probably this weekend. Thank you.

Reply:
Updated RFC to mention `foo ? null` pattern.